### PR TITLE
MON-2212: Expose the /federate endpoint of UWM Prometheus as a service

### PR DESCRIPTION
**Pull request by arajkumar**
Signed-off-by: Arunprasad Rajkumar <arajkuma@redhat.com>

* [X] I added CHANGELOG entry for this change.
* [ ] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

**Comment by arajkumar** [comment]
### Procedures to test

1. Launch a OCP cluster
2. Enable UWM and deploy an example app
```
oc apply -f https://gist.githubusercontent.com/arajkumar/7b3eaa8c21e6d74e8e0d7eaa7962af75/raw/8ff9dc4be8567183437e02170a245a1ee7a774ec/ocp-uwm.yaml
```
3. Bind cluster-monitoring-view RBAC to any service account or user to access `/federate` endpoint
```
oc apply -f https://gist.githubusercontent.com/arajkumar/605abfe388336b549b1555e82a6ba3f8/raw/56418ba20ee197b026048d0b887aeb77730c1279/federate-rbac.yaml
```
4. Access /federate endpoint
```
# get token
export TOKEN=$(oc sa get-token default -n ns1)

# port-forward
oc -n openshift-user-workload-monitoring port-forward service/prometheus-user-workload 9092:9092

curl -G -v -s -k -H "Authorization: Bearer ${TOKEN}" https://localhost:9092/federate --data-urlencode 'match[]=up'
```

**Comment by arajkumar** [comment]
/retest
/skip

Ping @juzhao for qe approval.
Ping @sferich888 for px approval.

**Comment by raptorsun** [comment]
Is it possible to share the same port & kube-rbac-proxy for metric and federate endpoints?
A suggestion:
- We change the --allow-path argument to --allow-path=/metrics,federate
- We include both RBAC settings in the config for kube-rbact-proxy as
```yaml
"authorization":
  "resourceAttributes":
    "apiVersion": "v1"
    "resource": "namespaces"
    "verb": "get"
  "static":
    - "path": "/metrics"
      "resourceRequest": false
      "user":
        "name": "system:serviceaccount:openshift-monitoring:prometheus-k8s"
      "verb": "get"

```

**Comment by arajkumar** [comment]
> Is it possible to share the same port & kube-rbac-proxy for metric and federate endpoints?

I don't think this would be good choice as the nature of access will be totally different. `/metrics` endpoint is meant for internal usage  but `/federate` called from outside the cluster by providing oauth token.

@simonpasquier @jan--f need you input on this. Thanks.

**Comment by jan--f** [comment]
> > Is it possible to share the same port & kube-rbac-proxy for metric and federate endpoints?
> 
> I don't think this would be good choice as the nature of access will be totally different. `/metrics` endpoint is meant for internal usage but `/federate` called from outside the cluster by providing oauth token.
> 
> @simonpasquier @jan--f need you input on this. Thanks.

Yes I agree with you @arajkumar. No user should be interested in the metrics the UWM prometheus exports.

File: pkg/manifests/manifests.go [hunk 1]
```diff
@@ -1796,7 +1808,8 @@ func (f *Factory) PrometheusUserWorkload(grpcTLS *v1.Secret) (*monv1.Prometheus,
 	}
 
 	for i, container := range p.Spec.Containers {
-		if container.Name == "kube-rbac-proxy" || container.Name == "kube-rbac-proxy-thanos" {
+		switch container.Name {
+		case "kube-rbac-proxy-metrics", "kube-rbac-proxy-federate", "kube-rbac-proxy-thanos":
```
Comment (jan--f) [inline pkg/manifests/manifests.go:1812]: Nice :)

**Review by jan--f — APPROVED** [review]
/lgtm

**Comment by juzhao** [comment]
@lihongyan1 please help to review

**Comment by arajkumar** [comment]
Ping @Senthamilarasu-STA (for px-approval)

**Pull request merged**